psr/with-let-letting.tex

```diff
@@ -0,0 +1,233 @@
+with certain exceptions defined in section
+\ref*{sec:let-variable-accessibility}.
+
+\subsubsection{Acessibility of variables defined using \gl{LET}}
+\label{sec:let-variable-accessibility}
+
+The \gl{GROUP BY} clause consumes the environments produced by the \gl{FROM} clause and
+produces a new set of environments, which effects the clauses which can access
+the variables defined by \gl{LET}.  Figure
+\ref*{figure:let-variable-accessibility} shows which parts of a \gl{SELECT}
+query can access variables defined by \gl{LET} with and without \gl{GROUP
+BY}.
+
+\begin{figure}[ht]
+\centering
+\begin{tabular}{lcc}    
+    \gl{SELECT} & \multicolumn{2}{c}{Are variables defined with \gl{LET} accessible?}  \\
+    clause & Without \gl{GROUP BY} & With \gl{GROUP BY}  \\
+    \hline
+    \gl{FROM} & No & No \\
+    \gl{LET} & Yes (per lexical scoping) & Yes (per lexical scoping) \\
+    \gl{WHERE} & Yes & Yes \\
+    \gl{GROUP BY} & N/A & Yes \\ 
+    \gl{LETTING} & N/A & No \\
+    \gl{HAVING} & N/A & No \\
+    \gl{ORDER BY} & Yes & No \\
+    \gl{OFFSET} & No & No  \\ 
+    \gl{LIMIT} & No & No \\
+    projection & Yes & No \\
+\end{tabular}
+\caption{Accessibility of variables defined using the \gl{LET} clause.}
+\label{figure:let-variable-accessibility}
+\end{figure}
+
+Note that variables defined by \gl{LET} are not accessible to the \gl{LIMIT} or
+\gl{OFFSET} clauses because those clauses do not have access to the environments
+produced by the \gl{FROM} or \gl{GROUP BY} clauses.
+
+
+\subsection{\gl{LETTING} clause}
+
+\gl{LETTING} is an optional clause of a \gl{SELECT} query that immediately
+follows the \gl{GROUP BY} clause. \gl{LETTING} executes once for every
+environment producted by the \gl{GROUP BY} clause, creating one nested
+environment for every \gn{var\_decl}. The \emph{last} environment produced by
+the \gl{LETTING} clause is accessible in the remaining clauses of the
+\gl{SELECT} query as defined in figure
+\ref*{figure:letting-variable-accessibility}.
+
+Use of \gl{LETTING} without \gl{GROUP BY} is considered to be an error.  
+
+\begin{figure}[ht]
+\centering
+\begin{tabular}{lcc}    
+    \gl{SELECT} & Are variables defined with \\
+    clause & \gl{LETTING} accessible? \\
+    \hline
+    \gl{FROM} & No \\
+    \gl{LET} & No \\
+    \gl{WHERE} & No \\
+    \gl{GROUP BY} & No \\ 
+    \gl{LETTING} & Yes (per lexical scoping) \\
+    \gl{HAVING} & Yes \\
+    \gl{ORDER BY} & Yes \\
+    \gl{OFFSET} & No  \\ 
+    \gl{LIMIT} & No \\
+    projection & Yes \\
+\end{tabular}
+\caption{Accessibility of variables defined using the \gl{LETTING} clause.}
+\label{figure:letting-variable-accessibility}
+\end{figure}
+
+
+\subsection{Behavior of \gl{WITH}, \gl{LET} and \gl{LETTING} and \gl{SELECT *}}
+
+\gl{SELECT *} is unaffected by the use of  \gl{WITH}, \gl{LET} and \gl{LETTING}.
+That is, variables introduced with these clauses are not included in
+\gl{SELECT *} projections.
```

Comment:
In the current implementation for `LET`, if there's a `GROUP BY` and a `LET` variable is included in the `GROUP BY`, those variables will be included in the projection.

For example with the query: ```SELECT * FROM `[{n: 10}, {n: 20}]` AS A LET 1 AS foo GROUP BY A.n, foo```

The result is:

```
<<
  {
    'n': 10,
    'foo': 1
  },
  {
    'n': 20,
    'foo': 1
  }
>>
```

If this is the intended behavior in the current implementation, then perhaps a better description would be something like: "`SELECT *` is unaffected ... unless the variable is included in the query's `GROUP BY`".

Reply:
I feel that this is *implicit* in the current definition.  I will consider spelling this out, however.